File: multi_roblox/launcher.py
"""Locate RobloxPlayerBeta.exe and spawn it with an auth ticket."""
import os
import subprocess
import time
from pathlib import Path
from typing import Optional

from . import auth

_CREATE_NO_WINDOW = 0x08000000


def find_player_exe() -> Optional[Path]:
    """Return the newest RobloxPlayerBeta.exe under %LOCALAPPDATA%\\Roblox\\Versions."""
    local = os.environ.get("LOCALAPPDATA")
    if not local:
        return None
    versions_dir = Path(local) / "Roblox" / "Versions"
    if not versions_dir.is_dir():
        return None
    candidates = []
    for child in versions_dir.iterdir():
        exe = child / "RobloxPlayerBeta.exe"
        if exe.is_file():
            candidates.append((exe.stat().st_mtime, exe))
    if not candidates:
        return None
    candidates.sort(reverse=True)
    return candidates[0][1]


def build_args(exe: Path, ticket: str, place_id: int,
               job_id: Optional[str] = None,
               browser_tracker_id: int = 0) -> list[str]:
    place_url = auth.place_launcher_url(place_id, job_id=job_id,
                                        browser_tracker_id=browser_tracker_id or None)
    return [
        str(exe),
        "--play",
        "-a", "https://www.roblox.com/Login/Negotiate.ashx",
        "-t", ticket,
        "-j", place_url,
        "-b", str(browser_tracker_id),
        f"--launchtime={auth.launchtime_ms()}",
        "--rloc", "en_us",
        "--gloc", "en_us",
    ]


def spawn(args: list[str]) -> subprocess.Popen:
    """Spawn the Roblox client detached so the manager outlives it cleanly."""
    return subprocess.Popen(
        args,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        stdin=subprocess.DEVNULL,
        creationflags=_CREATE_NO_WINDOW,
        close_fds=True,
    )


def launch_with_ticket(ticket: str, place_id: int,
                       job_id: Optional[str] = None) -> subprocess.Popen:
    exe = find_player_exe()
    if not exe:
        raise FileNotFoundError(
            "RobloxPlayerBeta.exe not found. Install Roblox and run it once."
        )
    # A loose browser tracker id, unique-ish per launch — not security sensitive.
    bti = int(time.time() * 1000) & 0x7FFFFFFF
    return spawn(build_args(exe, ticket, place_id, job_id=job_id, browser_tracker_id=bti))

File: csharp/MultiRobloxManager/Services/Launcher.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace MultiRobloxManager.Services;

internal static class Launcher
{
    private static FileInfo? NewestUnderVersions(string filename)
    {
        var versionsDir = Path.Combine(AppPaths.LocalAppData, "Roblox", "Versions");
        if (!Directory.Exists(versionsDir)) return null;
        FileInfo? newest = null;
        foreach (var dir in Directory.EnumerateDirectories(versionsDir))
        {
            var path = Path.Combine(dir, filename);
            if (!File.Exists(path)) continue;
            var fi = new FileInfo(path);
            if (newest is null || fi.LastWriteTimeUtc > newest.LastWriteTimeUtc) newest = fi;
        }
        return newest;
    }

    public static FileInfo? FindPlayerExe()   => NewestUnderVersions("RobloxPlayerBeta.exe");
    public static FileInfo? FindLauncherExe() => NewestUnderVersions("RobloxPlayerLauncher.exe");

    /// <summary>Returns e.g. "version-abc123" if discoverable, else null.</summary>
    public static string? DetectVersion()
    {
        var exe = FindPlayerExe();
        if (exe is null) return null;
        var parent = exe.Directory?.Name ?? "";
        return parent.StartsWith("version-", StringComparison.OrdinalIgnoreCase)
            ? parent.Substring("version-".Length)
            : parent;
    }

    /// <summary>Build the command-line args for direct RobloxPlayerBeta.exe exec with a ticket.</summary>
    public static List<string> BuildArgs(string ticket, long placeId, string? jobId, long browserTrackerId)
    {
        var placeUrl = Auth.PlaceLauncherUrl(placeId, jobId,
            browserTrackerId == 0 ? null : browserTrackerId);
        return new List<string>
        {
            "--play",
            "-a", "https://www.roblox.com/Login/Negotiate.ashx",
            "-t", ticket,
            "-j", placeUrl,
            "-b", browserTrackerId.ToString(),
            $"--launchtime={Auth.LaunchTimeMs()}",
            "--rloc", "en_us",
            "--gloc", "en_us",
        };
    }

    /// <summary>
    /// Spawn RobloxPlayerBeta.exe directly. Faster start, but skips the
    /// official launcher's update check and Hyperion parent-process setup.
    /// </summary>
    public static Process LaunchDirectWithTicket(string ticket, long placeId, string? jobId,
        IDictionary<string, string?>? extraEnv = null)
    {
        var exe = FindPlayerExe()
            ?? throw new FileNotFoundException(
                "RobloxPlayerBeta.exe not found. Install Roblox and run it once.");
        long bti = (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()) & 0x7FFFFFFF;
        var args = BuildArgs(ticket, placeId, jobId, bti);
        var psi = new ProcessStartInfo
        {
            FileName = exe.FullName,
            WorkingDirectory = exe.DirectoryName!,
            UseShellExecute = false,
            CreateNoWindow = true,
        };
        foreach (var a in args) psi.ArgumentList.Add(a);
        ApplyEnv(psi, extraEnv);
        var p = Process.Start(psi)!;
        Log.Info("spawn (direct): pid={0}", p.Id);
        return p;
    }

    /// <summary>
    /// Build the roblox-player: protocol URL with the auth ticket embedded
    /// as gameinfo. RobloxPlayerLauncher.exe handles updates and Hyperion's
    /// parent-process expectations.
    /// </summary>
    public static string ProtocolUrlWithTicket(string ticket, long placeId, string? jobId, long browserTrackerId)
    {
        var placeUrl = Auth.PlaceLauncherUrl(placeId, jobId,
            browserTrackerId == 0 ? null : browserTrackerId);
        // Percent-encode (Uri.EscapeDataString matches Python's
        // urllib.parse.quote default behavior); WebUtility.UrlEncode would
        // turn spaces into `+` which is form-encoding, not what we want here.
        var encodedPlace = Uri.EscapeDataString(placeUrl);
        var parts = new[]
        {
            "roblox-player:1",
            "launchmode:play",
            $"gameinfo:{ticket}",
            $"launchtime:{Auth.LaunchTimeMs()}",
            $"placelauncherurl:{encodedPlace}",
            $"browsertrackerid:{browserTrackerId}",
            "robloxLocale:en_us",
            "gameLocale:en_us",
        };
        return string.Join("+", parts);
    }

    /// <summary>
    /// Open the protocol URL via RobloxPlayerLauncher.exe so we can override
    /// env vars (LOCALAPPDATA for per-account isolation, when wired up).
    /// </summary>
    public static Process LaunchProtocolWithTicket(string ticket, long placeId, string? jobId,
        IDictionary<string, string?>? extraEnv = null)
    {
        long bti = (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()) & 0x7FFFFFFF;
        var url = ProtocolUrlWithTicket(ticket, placeId, jobId, bti);
        var launcherExe = FindLauncherExe();
        if (launcherExe is not null)
        {
            var psi = new ProcessStartInfo
            {
                FileName = launcherExe.FullName,
                WorkingDirectory = launcherExe.DirectoryName!,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            psi.ArgumentList.Add(url);
            ApplyEnv(psi, extraEnv);
            var p = Process.Start(psi)!;
            Log.Info("spawn (protocol via launcher): pid={0}", p.Id);
            return p;
        }
        // Fallback: hand off through the shell (no env override possible).
        var shellPsi = new ProcessStartInfo
        {
            FileName = url,
            UseShellExecute = true,
        };
        var sp = Process.Start(shellPsi)!;
        Log.Info("spawn (protocol via shell handoff): pid={0}", sp?.Id ?? -1);
        return sp!;
    }

    private static void ApplyEnv(ProcessStartInfo psi, IDictionary<string, string?>? extraEnv)
    {
        if (extraEnv is null) return;
        foreach (var (k, v) in extraEnv)
        {
            if (v is null) psi.Environment.Remove(k);
            else psi.Environment[k] = v;
        }
    }
}
